gc/gc-base/src/test/java/org/projectnessie/versioned/persist/gc/AbstractGCTest.java

```diff
@@ -0,0 +1,541 @@
+import org.projectnessie.client.http.NessieApiClient;
+import org.projectnessie.client.http.v1api.HttpApiV1;
+import org.projectnessie.model.CommitMeta;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.ImmutableCommitMeta;
+import org.projectnessie.model.Reference;
+import org.projectnessie.server.store.TableCommitMetaStoreWorker;
+import org.projectnessie.services.config.ServerConfig;
+import org.projectnessie.services.rest.RestContentResource;
+import org.projectnessie.services.rest.RestTreeResource;
+import org.projectnessie.versioned.BranchName;
+import org.projectnessie.versioned.GetNamedRefsParams;
+import org.projectnessie.versioned.Key;
+import org.projectnessie.versioned.NamedRef;
+import org.projectnessie.versioned.ReferenceAlreadyExistsException;
+import org.projectnessie.versioned.persist.adapter.ContentId;
+import org.projectnessie.versioned.persist.adapter.DatabaseAdapter;
+import org.projectnessie.versioned.persist.adapter.ImmutableCommitAttempt;
+import org.projectnessie.versioned.persist.adapter.KeyWithBytes;
+import org.projectnessie.versioned.persist.store.PersistVersionStore;
+import org.projectnessie.versioned.persist.tests.extension.DatabaseAdapterExtension;
+import org.projectnessie.versioned.persist.tests.extension.NessieDbAdapter;
+
+@ExtendWith(DatabaseAdapterExtension.class)
+abstract class AbstractGCTest {
+  @NessieDbAdapter protected static DatabaseAdapter databaseAdapter;
+
+  @ParameterizedTest
+  @MethodSource("datasetsSource")
+  void datasets(Dataset dataset) {
+    dataset.applyToAdapter(databaseAdapter);
+
+    datasetsTest(dataset, databaseAdapter);
+  }
+
+  static List<Dataset> datasetsSource() {
+    return produceDatasetsSource(Dataset::new);
+  }
+
+  static final TableCommitMetaStoreWorker STORE_WORKER = new TableCommitMetaStoreWorker();
+
+  static final String DEFAULT_BRANCH = "main";
+
+  // Note: since the GC implementation works with absolute timestamps, we can use absolute
+  // timestamps in our test data sets as well.
+
+  static final Instant NOW = Instant.ofEpochSecond(1000);
+  static final Instant NOT_LIVE = Instant.ofEpochSecond(500);
+
+  static ContentKey TABLE_ONE = ContentKey.of("table", "one");
+  static ContentKey TABLE_TWO = ContentKey.of("table", "two");
+  static ContentKey TABLE_THREE = ContentKey.of("table", "three");
+  static ContentKey TABLE_FOUR = ContentKey.of("table", "four");
+  static String CID_ONE = "table-one";
+  static String CID_TWO = "table-two";
+  static String CID_THREE = "table-three";
+
+  /**
+   * Common test implementation to verify the expected GC results for a specific data set. This test
+   * is exactly the same for mocked and "real" {@link DatabaseAdapter}s, datasets are exactly the
+   * same, timestamps are exactly the same, hashes however do vary.
+   */
+  protected void datasetsTest(Dataset dataset, DatabaseAdapter databaseAdapter) {
```

Comment:
This whole test framework is cryptic for the first timers to understand. 
But it is very scalable and easy to add test cases. Also validation part is very good and easy.
So, keeping this framework from @snazy's nessie-gc code.